File: drills/ts/solutions/nc113_course_schedule__dfs_colours.ts
export function canFinish(numCourses: number, prerequisites: number[][]): boolean {
  const needs: number[][] = Array.from({ length: numCourses }, () => []);
  for (const [course, prereq] of prerequisites) needs[course].push(prereq);

  const onPath = new Set<number>();
  const done = new Set<number>();

  // Depth-first with three states rather than two. "Seen" is not enough: a node
  // reached twice down *different* branches is fine, while a node reached again
  // while still on the current path is a cycle. The in-progress set is what
  // tells those apart.
  const visit = (course: number): boolean => {
    if (onPath.has(course)) return false;
    if (done.has(course)) return true;
    onPath.add(course);
    for (const prereq of needs[course]) {
      if (!visit(prereq)) return false;
    }
    onPath.delete(course);
    done.add(course);
    return true;
  };

  for (let course = 0; course < numCourses; course++) {
    if (!visit(course)) return false;
  }
  return true;
}

File: drills/go/solutions/nc113_course_schedule__dfs_colours.go
package main

func canFinish(numCourses int, prerequisites [][]int) bool {
	next := make([][]int, numCourses)
	for _, p := range prerequisites {
		next[p[1]] = append(next[p[1]], p[0])
	}
	// Three colours: unvisited, on the current path, finished. Reaching a
	// node on the current path is a cycle.
	const unvisited, onPath, done = 0, 1, 2
	colour := make([]int, numCourses)
	var acyclic func(course int) bool
	acyclic = func(course int) bool {
		switch colour[course] {
		case onPath:
			return false
		case done:
			return true
		}
		colour[course] = onPath
		for _, dependent := range next[course] {
			if !acyclic(dependent) {
				return false
			}
		}
		colour[course] = done
		return true
	}
	for course := 0; course < numCourses; course++ {
		if !acyclic(course) {
			return false
		}
	}
	return true
}
